template.yaml

```diff
@@ -527,6 +666,31 @@ Resources:
                 }],
                 "Default": "GetClusterSnapshotStatus"
               },
+              "CheckRegionSupportClusterSnapshotExportToS3": {
+                "Type": "Task",
+                "Resource": "${CheckRegionSupportClusterSnapshotExportToS3LambdaArn}",
+                "Next": "DoesRegionSupportClusterSnapshotExportToS3",
+                "ResultPath": "$"
+              },
+              "DoesRegionSupportClusterSnapshotExportToS3": {
+                "Type": "Choice",
+                "Choices": [{
+                  "Variable": "$.taskname",
+                  "StringEquals": "SnapshotExportSupported",
+                  "Next": "StartClusterSnapshotExportTask"
+                },
+                {
+                  "Variable": "$.taskname",
+                  "StringEquals": "Case of exporting snapshot to s3 in a region that supports snapshot export to s3.",
+                  "Next": "StartClusterSnapshotExportTask"
+                },
```

Comment:
@despot both the above choices lead to the StartSnapshotExportTask execution, can we check if the options be combined?
in the case of copy to specific region before export, can we check in the lambda itself regarding the branching/ exact task executed

Reply:
@namitad 
"both the above choices lead to the StartSnapshotExportTask execution, can we check if the options be combined?"
ok will combine the first two "StringEquals" in an "Or" condition.

"in the case of copy to specific region before export, can we check in the lambda itself regarding the branching/ exact task executed"
copy lambda does the copy from one region to another and takes the system to the next state GetDbSnapshotStatus. What do you mean by "branching/exact task executed" in copy lambda?

Reply:
@despot what i meant is, if we need to specify if it was directly exported from the source region or copied to a export_to_s3 supported region and then exported, this specification can be handled in the lambda if needed and 
the above 2 choices can be combined

Reply:
@namitad unfortunately it can't. 

If you checked the template.yaml, DbSnapshotCopy(step function of type Task) is parallel to StartSnapshotExportTask(step function of type Task). Before these two tasks come CheckRegionSupportSnapshotExportToS3(step function of type Task) and DoesRegionSupportSnapshotExportToS3(step function of type Choice). See the left most part of the attached image. 

And also consider that after DbSnapshotCopy the status(GetDbSnapshotStatus) and availability(isDbSnapshotAvailable) of the snapshot should be checked. See the upper part of the attached image. 

To answer you specific question, what you want to do is place DbSnapshotCopy between isDbSnapshotAvailable and StartSnapshotExportTask by including CheckRegionSupportSnapshotExportToS3 and DoesRegionSupportSnapshotExportToS3 in DbSnapshotCopy. This can't be done because of 3 reasons:
1. imho, the state machine should not pass copy of db snapshot (DbSnapshotCopy), if the region is supporting exporting (to reach StartSnapshotExportTask).
2. It becomes a router on top of a copy task (the states should do one job only - separation of concerns)
3. you **can't combine** an aws step function of **type="Task" and type="Choice"** at least [not from what I found online](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-choice-state.html).

![exportSnapshotToS3ForARegionThatDoesntSupportIt_useCase_detail](https://user-images.githubusercontent.com/1583742/96994501-1b42b400-152d-11eb-80af-0eedda1ebcaf.jpg)

If you agree with the above, kindly push the PR forward as it is waiting for 3 weeks.

If you disagree and you have a workable solution for this case, kindly lets meet and peer program (XP) as going back to the context of this PR is taking too much time for me (and I guess for you as well). Additionally I would suggest we wrap this last opened review issue, and not create new ones on the old code, as it takes time to reintegrate every new cosmetic fix and acceptance test it all (this last part takes money as well). For now, I would suggest, if the features work and you find another tech debt issue, to create a tech debt issue and accept / push the PR forward. In future, I would suggest doing a full review in order to avoid the time and money issues above. I am more than glad to help - don't get me wrong, but we have to be a bit pragmatic as well. I hope you understand. 
If there is anything else that you want to discuss before pushing the PR forward, kindly create a meeting call calendar invite (my email is despot dot jakimovski at gmail). Kindly, either place a link to the conference call or give me a skype hook or similar in the invite). I am available from UTC: 7am-10am or 5pm-8pm tomorrow(Saturday) and Sunday. Kindly schedule the call a day before preferably (at least half a day before).

Reply:
@despot i am not asking to combine `CheckRegionSupportSnapshotExportToS3 and DoesRegionSupportSnapshotExportToS3 in DbSnapshotCopy`
I would definitely want the `SnapshotExportSupported` and `Case of exporting snapshot to s3 in a region that supports snapshot export to s3.` choices to be merged.
I am merely wondering in the `StartClusterSnapshotExportTask` or the instance snapshot export task, there is a logging which specifies which if the snapshot was directly exported or copied and then exported (since you added 2 choices explicitly). But as i understand from the template.yml, the DbSnapshotCopy should give that info. Let me know if that understanding is correct.
Also, do add the updated state machine screenshot in the PR description because the number of additional steps is a lot more.

Reply:
@namitad 
"I would definitely want the SnapshotExportSupported and Case of exporting snapshot to s3 in a region that supports snapshot export to s3. choices to be merged."
This can't be done and I explained in my above comment why. And you are, kindly, not giving any explicit steps on how this can be performed with the existing limits.

"I am merely wondering in the StartClusterSnapshotExportTask or the instance snapshot export task, there is a logging which specifies which if the snapshot was directly exported or copied and then exported (since you added 2 choices explicitly)."
Start(Cluster)SnapshotExportTask tasks export from the appropriate region, which depends on the case constants.EXPORT_SNAPSHOT_TO_S3_IN_REGION_THAT_SUPPORTS_SNAPSHOT_EXPORT_TO_S3 , yes (check export_snapshot_s3_function.py).
"But as i understand from the template.yml, the DbSnapshotCopy should give that info."
No. The case constants.EXPORT_SNAPSHOT_TO_S3_IN_REGION_THAT_SUPPORTS_SNAPSHOT_EXPORT_TO_S3 is setup/given by the CheckRegionSupportSnapshotExportToS3 task (check check_region_support_snapshot_export_to_s3_function.py)
Kindly check the picture again.
Please setup a call if you still believe this can be changed. We kindly can't compress 10min of required conversation in few comments. It is taking too much time and it doesn't seem to be taking us to a feasible solution. 

As a final note, the case constants.EXPORT_SNAPSHOT_TO_S3_IN_REGION_THAT_SUPPORTS_SNAPSHOT_EXPORT_TO_S3 refers to the case when the default region is not supporting snapshot export to s3 so a copy was done to another region and the constant is propagated to notify the choice task whether to start an export from the appropriate region or copy it first. Again, kindly check the image.

"Also, do add the updated state machine screenshot in the PR description because the number of additional steps is a lot more."
Will do. Lets please first come to an understanding about the rest of this review issue.

Reply:
Hi @namitad 
cc: @stationeros 

if you don't schedule a call today or tomorrow, I will assume you understand and accept the limits, and I will fix:
1. [the default parameters issue](https://github.com/intuit/Trapheus/pull/34#issuecomment-715320778), and 
2. "add the updated state machine screenshot"
and acceptance test all of this in one go. After the push of this PR, I will expect that it will be accepted if the points 1. and 2. are working, as previously explained otherwise it will take a lot more time and money to go through another round of fixes and acceptance testing. 

Kind Regards,
Despot